gunrock/app/geo/geo_app.cu

```diff
@@ -0,0 +1,376 @@
+        "labels-file",
+        util::REQUIRED_ARGUMENT | util::SINGLE_VALUE | util::OPTIONAL_PARAMETER,
+        "",
+        "User locations label file for geolocation app.",
+        __FILE__, __LINE__));
+
+    return retval;
+}
+
+/**
+ * @brief Run geolocation tests
+ * @tparam     GraphT        Type of the graph
+ * @tparam     ValueT        Type of the distances
+ * @param[in]  parameters    Excution parameters
+ * @param[in]  graph         Input graph
+...
+ * @param[in]  target        where to perform the app 
+ * \return cudaError_t error message(s), if any
+ */
+template <typename GraphT>
+cudaError_t RunTests(
+    util::Parameters &parameters,
+    GraphT           &graph,
+    typename GraphT::ValueT *h_latitude,
+    typename GraphT::ValueT *h_longitude,
+    // <DONE> add problem specific reference results, e.g.:
+    typename GraphT::ValueT *ref_predicted_lat,
+    typename GraphT::ValueT *ref_predicted_lon,
+    // </DONE>
+    util::Location target)
+{
+    
+    cudaError_t retval = cudaSuccess;
+       
+    typedef typename GraphT::VertexT VertexT;
+    typedef typename GraphT::ValueT  ValueT;
+    typedef typename GraphT::SizeT   SizeT;
+    typedef Problem<GraphT>          ProblemT;
+    typedef Enactor<ProblemT>        EnactorT;
+
+    // CLI parameters
+    bool quiet_mode 		= parameters.Get<bool>("quiet");
+    int  num_runs   		= parameters.Get<int >("num-runs");
+    std::string validation 	= parameters.Get<std::string>("validation");
+
+    int geo_iter		= parameters.Get<int>("geo-iter");
+    util::PrintMsg("Number of iterations: " + geo_iter, !quiet_mode);
+
+    util::Info info("geolocation", parameters, graph);
+
+    util::CpuTimer cpu_timer, total_timer;
+    cpu_timer.Start(); total_timer.Start();
+
+    // <TODO> get problem specific inputs, e.g.:
+    // std::vector<VertexT> srcs = parameters.Get<std::vector<VertexT>>("srcs");
+    // printf("RunTests: %d srcs: src[0]=%d\n", srcs.size(), srcs[0]);
+    // </TODO>
+
+    // <DONE> allocate problem specific host data, e.g.:
+    ValueT *h_predicted_lat = new ValueT[graph.nodes];
+    ValueT *h_predicted_lon = new ValueT[graph.nodes];
+    // </DONE>
+
+    // Allocate problem and enactor on GPU, and initialize them
+    ProblemT problem(parameters);
+    EnactorT enactor;
+
+    util::PrintMsg("Initializing problem ... ", !quiet_mode);
+
+    GUARD_CU(problem.Init(graph, target));
+
+    util::PrintMsg("Initializing enactor ... ", !quiet_mode);
+
+    GUARD_CU(enactor.Init(problem, target));
+    
+    cpu_timer.Stop();
+    parameters.Set("preprocess-time", cpu_timer.ElapsedMillis());
+    
+    for (int run_num = 0; run_num < num_runs; ++run_num) {
+        GUARD_CU(problem.Reset(
+            // <DONE> problem specific data if necessary, eg:
+            h_latitude,
+	    h_longitude,
+	    geo_iter,
+            // </DONE>
+            target
+        ));
+        GUARD_CU(enactor.Reset(
+            // <TODO> problem specific data if necessary:
```

Comment:
Probably a good idea to remove complete `<TODO>`'s, so people looking at the code don't think that this is an incomplete implementation.

Reply:
Done. (maybe I missed some)